src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/MarshallerShape.cs

```diff
@@ -174,4 +209,210 @@ static bool IsSpanOfUnmanagedType(ITypeSymbol typeToCheck, ITypeSymbol spanOfT,
                     && SymbolEqualityComparer.Default.Equals(managedType, m.ReturnType));
         }
     }
+
+    public static class StatefulMarshallerShapeHelper
+    {
+        public record MarshallerMethods
+        {
+            public IMethodSymbol? FromManaged { get; init; }
+            public IMethodSymbol? FromManagedWithBuffer { get; init; }
+            public IMethodSymbol? ToManaged { get; init; }
+            public IMethodSymbol? ToManagedGuranteed { get; init; }
+            public IMethodSymbol? FromUnmanaged { get; init; }
+            public IMethodSymbol? ToUnmanaged { get; init; }
+            public IMethodSymbol? Free { get; init; }
+            public IMethodSymbol? NotifyInvokeSucceeded { get; init; }
+        }
+
+        public static (MarshallerShape shape, MarshallerMethods methods) GetShapeForType(ITypeSymbol marshallerType, ITypeSymbol managedType, Compilation compilation)
+        {
+            MarshallerShape shape = MarshallerShape.None;
+            MarshallerMethods methods = new();
+
+            ITypeSymbol? unmanagedType = null;
+
+            IMethodSymbol? fromManaged = GetFromManagedMethod(marshallerType, managedType);
+            INamedTypeSymbol spanOfT = compilation.GetTypeByMetadataName(TypeNames.System_Span_Metadata)!;
+            IMethodSymbol? fromManagedWithCallerAllocatedBuffer = GetFromManagedWithCallerAllocatedBufferMethod(marshallerType, managedType, spanOfT, out _);
+
+            IMethodSymbol? toUnmanaged = GetToUnmanagedMethod(marshallerType);
+
+            if ((fromManaged, fromManagedWithCallerAllocatedBuffer) is not (null, null) && toUnmanaged is not null)
+            {
+                unmanagedType = toUnmanaged.ReturnType;
+                if (unmanagedType.IsUnmanagedType)
+                {
+                    if (fromManagedWithCallerAllocatedBuffer is not null)
+                    {
+                        shape |= MarshallerShape.CallerAllocatedBuffer;
+                    }
+                    if (fromManaged is not null)
+                    {
+                        shape |= MarshallerShape.ToUnmanaged;
+                    }
+                    methods = methods with
+                    {
+                        FromManaged = fromManaged,
+                        FromManagedWithBuffer = fromManagedWithCallerAllocatedBuffer,
+                        ToUnmanaged = toUnmanaged
+                    };
+                }
+            }
+
+            IMethodSymbol toManaged = GetToManagedMethod(marshallerType, managedType);
+            IMethodSymbol toManagedGuaranteed = GetToManagedGuaranteedMethod(marshallerType, managedType);
+            ImmutableArray<IMethodSymbol> fromUnmanagedCandidates = ImmutableArray.CreateRange(GetFromUnmanagedMethodCandidates(marshallerType));
```

Comment:
If we allow multiple overloads of the `FromUnmanaged` method, then we can end up in a weird state where we pick the first method in whatever order `GetMembers` returns that meets our criteria, which is not well defined when you have a `partial` type (though it is consistent for a given Roslyn command line invocation).

Additionally, it felt weird to support multiple overloads of `FromUnmanaged` as long as one matched the return type of `ToUnmanaged` since an unmanaged->managed-only marshaller wouldn't be allowed to have overloads, but a bidirectional marshaller would.

We can move the check of the candidates into the method and only return a non-`null` `IMethodSymbol` if we have only one result and that result is valid if you'd prefer

Reply:
Ah, got it - and `ToUnanaged` is fine since you can't have multiple with just different return types, right? I'd prefer moving it into the method and having a comment there about why we do the check.

Reply:
I'll move it into the method then